codepropertygraph/src/main/scala/io/shiftleft/codepropertygraph/cpgloading/CpgOverlayLoader.scala

```diff
@@ -46,17 +48,35 @@ private[cpgloading] object CpgOverlayLoader {
 private class CpgOverlayApplier(graph: ScalaGraph) {
 
   private val overlayNodeIdToSrcGraphNode: mutable.HashMap[Long, Vertex] = mutable.HashMap()
+  private var inverseBuilder: DiffGraph.InverseBuilder = _
 
   /**
     * Applies diff to existing (loaded) OdbGraph
     */
   def applyDiff(overlay: CpgOverlay): Unit = {
+    inverseBuilder = makeInverseBuilder(false)
     addNodes(overlay)
     addEdges(overlay)
     addNodeProperties(overlay)
     addEdgeProperties(overlay)
   }
 
+  def applyUndoableDiff(overlay: CpgOverlay): DiffGraph = {
+    inverseBuilder = makeInverseBuilder(true)
+    addNodes(overlay)
+    addEdges(overlay)
+    addNodeProperties(overlay)
+    addEdgeProperties(overlay)
+    inverseBuilder.build()
+  }
+
+  private def makeInverseBuilder(undoable: Boolean) = {
+    if (undoable)
+      DiffGraph.InverseBuilder.newBuilder
+    else
+      DiffGraph.InverseBuilder.noop
+  }
```

Comment:
I would argue that this could be removed completely and inlined in the corresponding functions

Reply:
It could but it would make 8+ duplicated if statements

Reply:
@GlassAndOneHalf see also https://github.com/ShiftLeftSecurity/codepropertygraph/pull/552#discussion_r368772969